File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        int count = 0, sum =0;
        unordered_map<int,int> prefix;
        prefix[0] = 1;
        for(int i=0; i<n ; i++){
            sum+=nums[i];

            if(prefix.find(sum-k) != prefix.end()){
                count+=prefix[sum-k];
            }
            prefix[sum]++;
        }
        return count;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int n = nums.length;
        int count = 0, sum =0;
        HashMap<Integer,Integer> freq = new HashMap<>(); // prefix;
        freq.put(0,1);
        for(int i=0; i<n ; i++){
            sum+=nums[i];

            if(freq.containsKey(sum-k)){
                count+=freq.get(sum-k);
            }
            freq.put(sum,freq.getOrDefault(sum,0)+1);
        }
        return count;
    }
}